spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/connection/TcpNetConnection.java

```diff
@@ -234,43 +262,32 @@ protected boolean handleReadException(Exception e) {
 				logger.error("Error accessing soTimeout", e1);
 			}
 		}
-		if (doClose) {
-			boolean noReadErrorOnClose = isNoReadErrorOnClose();
-			closeConnection(true);
-			if (!(e instanceof SoftEndOfStreamException)) {
-				if (e instanceof SocketTimeoutException) {
-					if (logger.isDebugEnabled()) {
-						logger.debug("Closed socket after timeout:" + getConnectionId());
-					}
-				}
-				else {
-					if (noReadErrorOnClose) {
-						if (logger.isTraceEnabled()) {
-							logger.trace("Read exception " +
-									getConnectionId(), e);
-						}
-						else if (logger.isDebugEnabled()) {
-							logger.debug("Read exception " +
-									getConnectionId() + " " +
-									e.getClass().getSimpleName() +
-									":" + (e.getCause() != null ? e.getCause() + ":" : "") + e.getMessage());
-						}
-					}
-					else if (logger.isTraceEnabled()) {
-						logger.error("Read exception " +
-								getConnectionId(), e);
-					}
-					else {
-						logger.error("Read exception " +
-								getConnectionId() + " " +
-								e.getClass().getSimpleName() +
-								":" + (e.getCause() != null ? e.getCause() + ":" : "") + e.getMessage());
-					}
-				}
-				sendExceptionToListener(e);
+		return doClose;
+	}
+
+	private void logOtherExceptions(Exception e, boolean noReadErrorOnClose) {
+		if (noReadErrorOnClose) {
+			if (logger.isTraceEnabled()) {
+				logger.trace("Read exception " +
+						getConnectionId(), e);
+			}
+			else if (logger.isDebugEnabled()) {
+				logger.debug("Read exception " +
+						getConnectionId() + " " +
+						e.getClass().getSimpleName() +
+						":" + (e.getCause() != null ? e.getCause() + ":" : "") + e.getMessage());
 			}
 		}
-		return doClose;
+		else if (logger.isTraceEnabled()) {
+			logger.error("Read exception " +
```

Comment:
It is intentional to use `error()` here when you check for `trace`?
Can we have this whole method reworked somehow to avoid duplication?

Thanks